src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingSourceGenerator.cs

```diff
@@ -0,0 +1,182 @@
+
+//#define LAUNCH_DEBUGGER
+using System.Collections.Immutable;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.DotnetRuntime.Extensions;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.Extensions.Configuration.Binder.SourceGeneration
+{
+    /// <summary>
+    /// Generates source code to optimize binding with ConfigurationBinder.
+    /// </summary>
+    [Generator]
+    public sealed partial class ConfigurationBindingSourceGenerator : IIncrementalGenerator
+    {
+        public void Initialize(IncrementalGeneratorInitializationContext context)
+        {
+            IncrementalValueProvider<KnownTypeData> compilationData =
+                context.CompilationProvider
+                    .Select((compilation, _) => new KnownTypeData(compilation));
+
+            IncrementalValuesProvider<BinderInvocationOperation> inputCalls = context.SyntaxProvider.CreateSyntaxProvider(
+                (node, _) => node is InvocationExpressionSyntax invocation,
+                (context, cancellationToken) => new BinderInvocationOperation(context, cancellationToken));
+
+            IncrementalValueProvider<(KnownTypeData, ImmutableArray<BinderInvocationOperation>)> inputData = compilationData.Combine(inputCalls.Collect());
+
+            context.RegisterSourceOutput(inputData, (spc, source) => Execute(source.Item1, source.Item2, spc));
+        }
+
+        /// <summary>
+        /// Generates source code to optimize binding with ConfigurationBinder.
+        /// </summary>
+        private static void Execute(KnownTypeData typeData, ImmutableArray<BinderInvocationOperation> inputCalls, SourceProductionContext context)
+        {
+#if LAUNCH_DEBUGGER
+            if (!System.Diagnostics.Debugger.IsAttached)
+            {
+                System.Diagnostics.Debugger.Launch();
+            }
+#endif
+            if (inputCalls.IsDefaultOrEmpty)
+            {
+                return;
+            }
+
+            Parser parser = new(context, typeData);
+            SourceGenerationSpec? spec = parser.GetSourceGenerationSpec(inputCalls);
+            if (spec is not null)
+            {
+                Emitter emitter = new(context, spec);
+                emitter.Emit();
+            }
+        }
+
+        private sealed record KnownTypeData
+        {
+            public INamedTypeSymbol SymbolForGenericIList { get; }
+            public INamedTypeSymbol SymbolForICollection { get; }
+            public INamedTypeSymbol SymbolForIEnumerable { get; }
+            public INamedTypeSymbol SymbolForString { get; }
+
+            public INamedTypeSymbol? SymbolForConfigurationKeyNameAttribute { get; }
+            public INamedTypeSymbol? SymbolForDictionary { get; }
+            public INamedTypeSymbol? SymbolForGenericIDictionary { get; }
+            public INamedTypeSymbol? SymbolForHashSet { get; }
```

Comment:
>SymbolForHashSet [](http://example.com/codeflow?start=37&length=16)

I am seeing you used this in the parser but you never emit HashSet (and ISet) maybe others too. 

Should I assume this is something you'll support later?

Reply:
Those types are supported; just need to follow up with readonly collections shortly - https://github.com/dotnet/runtime/issues/79527.